File: NapakalakiJava/src/NapakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.*;
/**
 *
 * @author danibolanos & jomabose
 */
public class Napakalaki {
    
    private static final Napakalaki instance = new Napakalaki();
    
    private Monster currentMonster;
    private CardDealer dealer;
    private Player currentPlayer;
    private ArrayList<Player> players;
    
    private Napakalaki() {
        //currentMonster
        //dealer
        currentPlayer = null;
        players = new ArrayList();
    }
    
    private void initPlayers(ArrayList<String> name){
        
        for(int i=0; i < name.size(); ++i)
           players.add(new Player(name.get(i)));
    }
    
    private Player nextPlayer(){
        int indice;
        
        if (currentPlayer == null){
           indice = (int) (Math.random()*players.size());
           currentPlayer = players.get(indice);
        }
        
        else{
            indice = players.indexOf(currentPlayer)+1;
            
            if(indice==players.size())
                currentPlayer = players.get(0);
            
            else
                currentPlayer = players.get(indice);
        }
        
        return currentPlayer;
    }
    
    private boolean nextTurnAllowed(){
        boolean permitido = false;
        
        if (currentPlayer == null)
            permitido = true;
        
        else{
            
            if(currentPlayer.validState())
                permitido = true;
        }
        
        return permitido;
    }
    
    private void setEnemies(){
        int indice;
        
        for(int i=0; i<players.size(); i++){
            indice = (int) (Math.random()*players.size()-1);
            
            if(indice==i)
                indice = players.size()-1;
            players.get(i).setEnemy(players.get(indice));
        }
    }
    
    public static Napakalaki getInstance() {
        return instance;
    }
    
    public CombatResult developCombat(){
        CombatResult combatResult = currentPlayer.combat(currentMonster);
        
        if(combatResult==CombatResult.LOSEANDCONVERT){
           CultistPlayer sectario = new CultistPlayer(currentPlayer, CardDealer.getInstance().nextCultist());
           
           for(Player jugador:players){
               
               if(jugador.getEnemy()==currentPlayer)
                   jugador.setEnemy(sectario);
           }  
           
           int indiceCurrentPlayer = players.indexOf(currentPlayer);
           players.set(indiceCurrentPlayer, sectario);
           currentPlayer = sectario;
        }
        CardDealer.getInstance().giveMonsterBack(currentMonster);
        
        return combatResult;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        
        for(Treasure treasure:treasures){
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        
        for(Treasure treasure:treasures){
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        
        for(Treasure t:treasures){
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public void initGame(ArrayList<String> players){
        initPlayers(players);
        setEnemies();
        dealer = CardDealer.getInstance();
        dealer.initCards();
        nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean stateOK = nextTurnAllowed();
        
        if(stateOK){
            currentMonster = dealer.nextMonster();
            currentPlayer = nextPlayer();
            boolean dead = currentPlayer.isDead();
            
            if(dead)
                currentPlayer.initTreasures();
        }
        
        return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        boolean fin=false;
        
        if(result == CombatResult.WINGAME)
            fin=true;
        
        return fin;
    }
    
    public String toString(){
        String cadena = "Monstruo actual: " + currentMonster.toString();
        cadena += "\nJugador actual: " + currentPlayer.toString();
        
        cadena += "\nTodos los jugadores: ";
        for(int i=0; i<players.size(); i++)
            cadena += "\n" + players.get(i).toString();
        cadena += "\nBaraja: " + dealer.toString();
        
        return cadena;
    }
}

File: EXAMENES_PDOO/NapakalakiRuby/lib/napakalaki.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame

require 'singleton'
require_relative 'player'
require_relative 'card_dealer'
require_relative 'monster'
require_relative 'combat_result'

class Napakalaki
  include Singleton
  
  private
  
  def initialize
    @currentMonster = nil
    @currentPlayer = nil
    @players = nil
    @dealer = nil
  end

  def initPlayers(names)
    @players = Array.new
    names.each do |nombre|
       @players << Player.new(nombre)
    end
  end
  
  def nextPlayer
    index = 0
    if @currentPlayer == nil then
       index = rand(@players.size)
    else
      index = @players.index(@currentPlayer) + 1
      if index == @players.size then
        index = 0        
      end
    end
    @currentPlayer = @players.at(index)
    @players.at(index)
  end
  
  def nextTurnAllowed
    condicion = false
    if @currentPlayer == nil || @currentPlayer.validState then
      condicion = true
    end
    condicion
  end
  
  def setEnemies
    @players.each do |jugador|
      enemigo = rand((@players.size)-1)
      if enemigo == @players.index(jugador) then
       enemigo = (@players.size)-1
      end
       jugador.setEnemy(@players.at(enemigo))
    end
  end
  
  public
  
  def developCombat
    combatResult = @currentPlayer.combat(@currentMonster)
    @dealer.giveMonsterBack(@currentMonster)
    combatResult
  end
  
  def discardVisibleTreasures(treasures)
    treasures.each do |treasure|
      @currentPlayer.discardVisibleTreasure(treasure)
      @dealer.giveTreasureBack(treasure)
    end
  end
  
  def discardHiddenTreasures(treasures)
    treasures.each do |treasure|
      @currentPlayer.discardHiddenTreasure(treasure)
      @dealer.giveTreasureBack(treasure)
    end
  end
  
  def makeTreasuresVisible(treasures)
    treasures.each do |t|
      @currentPlayer.makeTreasureVisible(t)
    end
  end
  
  def initGame(players)
    initPlayers(players)
    setEnemies
    @dealer = CardDealer.instance
    @dealer.initCards
    nextTurn
  end
  
  def getCurrentPlayer
    @currentPlayer
  end
  
  def getCurrentMonster
    @currentMonster
  end
  
  def nextTurn
    stateOK = nextTurnAllowed
    if stateOK then
      @currentMonster = @dealer.nextMonster
      @currentPlayer = nextPlayer
      dead = @currentPlayer.isDead
      if dead then
        @currentPlayer.initTreasures
      end
    end
    stateOK
  end
  
  def endOfGame(result)
    condicion = false
    if result == [CombatResult::WINGAME] then
      condicion = true
    end
    condicion
  end 
end

end